File: src/main/java/com/fwdekker/randomness/string/Alphabet.java
package com.fwdekker.randomness.string;

import java.util.Collection;
import java.util.stream.Collectors;


/**
 * An {@code Alphabet} represents a collection of symbols.
 */
enum Alphabet {
    UPPERCASE("Uppercase (A, B, C, ...)", "ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
    LOWERCASE("Lowercase (a, b, c, ...)", "abcdefghijklmnopqrstuvwxyz"),
    DIGITS("Digits (0, 1, 2, ...)", "0123456789"),
    MINUS("Minus (-)", "-"),
    UNDERSCORE("Underscore (_)", "_"),
    SPACE("Space ( )", " "),
    SPECIAL("Special (!, @, #, $, %, ^, &, *)", "!@#$%^&*"),
    BRACKETS("Brackets ((, ), [, ], {, }, <, >)", "()[]{}<>");


    /**
     * The name of the alphabet.
     */
    private final String name;
    /**
     * The symbols in the alphabet.
     */
    private final String symbols;


    /**
     * Constructs a new {@code Alphabet}.
     *
     * @param name    the name of the alphabet
     * @param symbols the symbols in the alphabet
     */
    Alphabet(final String name, final String symbols) {
        this.name = name;
        this.symbols = symbols;
    }


    /**
     * Concatenates the symbols of all the alphabets in the given collection.
     *
     * @param alphabets a collection of alphabets
     * @return the concatenation of all symbols of all the alphabets in the given collection
     */
    static String concatenate(final Collection<Alphabet> alphabets) {
        return alphabets.stream()
                .map(Alphabet::getSymbols)
                .collect(Collectors.joining());
    }


    /**
     * Returns the name of the alphabet.
     *
     * @return the name of the alphabet
     */
    String getName() {
        return name;
    }

    /**
     * Returns the symbols in the alphabet.
     *
     * @return the symbols in the alphabet
     */
    String getSymbols() {
        return symbols;
    }

    @Override
    public String toString() {
        return name;
    }
}

File: src/main/kotlin/com/fwdekker/randomness/string/Alphabet.kt
package com.fwdekker.randomness.string


/**
 * An `Alphabet` represents a named collection of symbols.
 *
 * @param description the description of the alphabet
 * @param symbols the symbols in the alphabet
 */
data class Alphabet(val description: String, val symbols: String) {
    companion object {
        /**
         * The lowercase English alphabet.
         */
        val ALPHABET = Alphabet("Alphabet (a, b, c, ...)", "abcdefghijklmnopqrstuvwxyz")
        /**
         * The digits 0 through 9.
         */
        val DIGITS = Alphabet("Digits (0, 1, 2, ...)", "0123456789")
        /**
         * The hexadecimal digits 0 through f.
         */
        val HEXADECIMAL = Alphabet("Hexadecimal (0, 1, 2, ..., d, e, f)", "0123456789abcdef")
        /**
         * A minus (`-`).
         */
        val MINUS = Alphabet("Minus (-)", "-")
        /**
         * An underscore (`_`).
         */
        val UNDERSCORE = Alphabet("Underscore (_)", "_")
        /**
         * A whitespace (` `).
         */
        val SPACE = Alphabet("Space ( )", " ")
        /**
         * A collection of special characters.
         */
        val SPECIAL = Alphabet("Special (!, @, #, $, %, ^, &, *)", "!@#$%^&*")
        /**
         * A collection of brackets and parentheses.
         */
        val BRACKETS = Alphabet("Brackets ((, ), [, ], {, }, <, >)", "()[]{}<>")

        /**
         * List of default alphabets.
         */
        val defaultAlphabets = listOf(ALPHABET, DIGITS, HEXADECIMAL, MINUS, UNDERSCORE, SPACE, SPECIAL, BRACKETS)
    }


    /**
     * Returns the `description` field.
     *
     * @return the `description` field
     */
    override fun toString() = description
}


/**
 * Concatenates the symbols of all the alphabets, removing duplicate characters.
 *
 * @return the concatenation of all symbols of all the alphabets, excluding duplicate characters
 */
fun Iterable<Alphabet>.sum() =
    this.fold("") { acc, alphabet -> acc + (alphabet.symbols.takeUnless { it in acc } ?: "") }
